lib/std/os/windows.zig

```diff
@@ -937,25 +937,50 @@ pub fn DeleteFile(sub_path_w: []const u16, options: DeleteFileOptions) DeleteFil
         .DELETE_PENDING => return,
         else => return unexpectedStatus(rc),
     }
-    var file_dispo = FILE_DISPOSITION_INFORMATION{
-        .DeleteFile = TRUE,
-    };
-    rc = ntdll.NtSetInformationFile(
-        tmp_handle,
-        &io,
-        &file_dispo,
-        @sizeOf(FILE_DISPOSITION_INFORMATION),
-        .FileDispositionInformation,
-    );
-    CloseHandle(tmp_handle);
-    switch (rc) {
-        .SUCCESS => return,
-        .DIRECTORY_NOT_EMPTY => return error.DirNotEmpty,
-        .INVALID_PARAMETER => unreachable,
-        .CANNOT_DELETE => return error.AccessDenied,
-        .MEDIA_WRITE_PROTECTED => return error.AccessDenied,
-        .ACCESS_DENIED => return error.AccessDenied,
-        else => return unexpectedStatus(rc),
+    defer CloseHandle(tmp_handle);
+    if (comptime builtin.target.os.version_range.windows.min.isAtLeast(.win10_rs1)) {
+        // Deletion with posix semantics.
+        var info = FILE_DISPOSITION_INFORMATION_EX{
+            .Flags = FILE_DISPOSITION_DELETE |
+                FILE_DISPOSITION_POSIX_SEMANTICS |
+                FILE_DISPOSITION_ON_CLOSE |
+                FILE_DISPOSITION_IGNORE_READONLY_ATTRIBUTE,
+        };
+
+        rc = ntdll.NtSetInformationFile(
+            tmp_handle,
+            &io,
+            &info,
+            @sizeOf(FILE_DISPOSITION_INFORMATION_EX),
+            .FileDispositionInformationEx,
+        );
+        switch (rc) {
+            .SUCCESS => {},
+            .CANNOT_DELETE => return error.FileBusy, // file is currently mapped
+            else => return unexpectedStatus(rc),
```

Comment:
Some notes / questions:

- `CANNOT_DELETE` is converted to `error.FileBusy` here but `error.AccessDenied` in the branch below. They should be converted to the same error.
- All the error conditions within the branch below should probably be handled here too (see https://github.com/ziglang/zig/pull/15316#discussion_r1169355973)
- Need to figure out exactly how `FILE_DISPOSITION_IGNORE_READONLY_ATTRIBUTE` affects things and update [https://github.com/ziglang/zig/blob/440b3df702f1c8bfddabc1c594a3f49cf0011a63/lib/std/fs/test.zig#L1419-L1436](this test) accordingly.

Reply:
All comments were addressed by using the same error return set and adjusting the test: Read-only flags for files are ignored upon deletion and the tests were adjusted to take that into account. Once EOL for win11 is reached, the test block with "pending file deletion semantics" can be deleted.